docs/guide/custom-login-requirements.md

```diff
@@ -32,15 +32,112 @@ While [most password complexity rules are often counterproductive](https://arste
+
+Each requirement has:
+
+* A `phase` property, which must be `beforeSubmit`, `afterSubmit` or `afterPasswordVerified`.
+* A `props(req)` function, which may be `async` and returns an object. The properties of that object become Vue `props` of the Vue component, as explained below. The `req` object allows access to the [current Express request](https://expressjs.com/en/api.html#req), including `req.session` which can be useful for temporary storage not revealed to the browser.
+* A `verify(req)` function, which may be `async`. This function is responsible for throwing an exception if the requirement has not been met. Any data provided by the Vue component will be accessible here as `req.body.requirements.RequirementName`, where `RequirementName` should be replaced with the name of the requirement.
+
+To illustrate, the general structure on the server side is:
+
+<AposCodeBlock>
+  ```javascript
+  module.exports = {
+    requirements(self) {
+      return {
+        add: {
+          RequirementName: {
+            phase: 'afterSubmit',
+            async props(req) {
+              return {
+                propName: propValue
+              };
+            },
+            async verify(req) {
+              if (!verifyThisCodeIsCorrect(req.body.requirement.RequirementName)) {
+                throw self.apos.error('invalid', 'appropriate message');
+              }
+            }
+          }
+        }
+      };
+    }
+  };
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/login/index.js
+  </template>
+</AposCodeBlock>
+
+## The browser side
+
+Each distinct requirement also has browser-side code, which is implemented as a Vue component. As explained in the [custom UI guide](../custom-ui.md), Vue components intended for the admin UI (including login requirements) must be placed in the `ui/apos/components` subdirectory of a module in the project. For this purpose they are typically placed in the `modules/@apostrophecms/login/ui/apos/components` module at project level, or in an npm module that enhances `@apostrophecms/login` via `improve`.
+
+The developer is responsible for the appearance of the component. For `beforeSubmit` requirements, the component will appear at the bottom of the login form itself. For `afterSubmit` requirements, it will appear on its own, "fading up" after the Login button is clicked, although password verification has not happened yet. For `afterPasswordVerified` requirements, the appearance is the same as for `afterSubmit`, although password verification has been completed at this point. If there are multiple `afterSubmit` and `afterPasswordVerified` requirements, the user will see them presented one at a time, with the `afterSubmit` requirements completed first.
+
+Each component is responsible for:
+
+* Presenting its own UI, if any. Apostrophe will handle "fading up" `afterSubmit` and `afterPasswordVerified` components and displaying any suitable outer framing.
```

Comment:
Same regarding "fading up."

Reply:
We'll want to provide clearer direction here regarding how the outer framing is applied if they need to include that in their components. We don't have that information yet, though.